File: practice/algorithms/dynamic_programming/max_subarray/MaxSubarray.scala
import scala.io.Source

object MaxSubarray extends App {
  val lines = Source.stdin.getLines.drop(1).toList

  val arrayLines = lines.grouped(2).map(group => group.toList(1))
  val arrays = arrayLines.map(_.split(" ").map(_.toInt))
  arrays.foreach {
    arr => {
      val contiguousSum = kadane(arr)
      val positiveNumbers = arr.filter(_ > 0)
      val nonContiguousSum = if(positiveNumbers.isEmpty) arr.max else positiveNumbers.sum
      println(contiguousSum + " " + nonContiguousSum)
    }
  }

  def kadane(arr: Array[Int]): Int = {
    var maxSoFar = arr(0)
    var maxEndings = arr(0)
    arr.slice(1, arr.length).foreach(el => {
      maxEndings = scala.math.max(el, maxEndings + el)
      maxSoFar = scala.math.max(maxSoFar, maxEndings)
    })
    maxSoFar
  }
}

File: practice/algorithms/dynamic_programming/max_subarray/MaxSubarray.java
import java.util.*;

public class MaxSubarray {

    public static void main(String[] args) {
        Scanner stdin = new Scanner(System.in);
        int tests = stdin.nextInt();
        for(int i = 0; i < tests; i++) {
            int elementCount = stdin.nextInt();
            int[] arr = new int[elementCount];
            for(int j = 0; j < elementCount; j++) {
                arr[j] = stdin.nextInt();
            }
            int contiguousSum = kadane(arr);
            int nonContiguousSum;
            if(hasPositiveElement(arr)) {
                nonContiguousSum = sumPositiveElement(arr);
            } else {
                nonContiguousSum = max(arr);
            }
            System.out.println(contiguousSum + " " + nonContiguousSum);
        }
        stdin.close();
    }

    private static int kadane(int[] arr) {
        int maxSoFar = arr[0];
        int maxEndings = arr[0];
        for(int j = 1; j < arr.length; j++) {
            int el = arr[j];
            maxEndings = Math.max(el, maxEndings + el);
            maxSoFar = Math.max(maxSoFar, maxEndings);
        }
        return maxSoFar;
    }

    private static boolean hasPositiveElement(int[] arr) {
        for(int j = 0; j < arr.length; j++) {
            if(arr[j] > 0) {
                return true;
            }
        }
        return false;
    }

    private static int sumPositiveElement(int[] arr) {
        int sum = 0;
        for(int j = 0; j < arr.length; j++) {
            if(arr[j] > 0) {
                sum += arr[j];
            }
        }
        return sum;
    }
    private static int max(int[] arr) {
        int max = Integer.MIN_VALUE;
        for(int j = 0; j < arr.length; j++) {
            if(arr[j] > max) {
                max = arr[j];
            }
        }
        return max;
    }
}